### ISSUE TITLE
`xontrib` tutorial should be tweaked to clarify more details, how to not do autoloading.. etc.

### ISSUE THREAD

**Issue by goodboy** [xonsh/xonsh#5125]
This took me a shamefully long time to understand, mostly due to the combination of the docs being organized somewhat *in a strange order* (in my humble opinion) as well as reading through a bunch of `awesome-rc-xxx` repos and the [template repo](https://github.com/xonsh/xontrib-cookiecutter) which also did not clarify what I hopefully now understand 😂 

Actually, reading other xontrib projects, the core `xonsh` source code and reversing engineering how other repos do things ended up being more helpful then the actual docs, which imo is not that ideal 😂 

So hopefully this issue serves to at least one of either, help others who were as confused about the details as I, and provide motivation to make the docs a bit more detailed (with potentially a more in depth example) to explain the multiple ways in which `xontrib`s can be written and registered for use.

As an example of a project I work on on which eventually offered something similar, check out the [complete example](https://pluggy.readthedocs.io/en/latest/#a-complete-example) section of the `pluggy` docs (the plugin sys for `pytest`).

To begin i'm going to summarize what i think i do and don't understand.. and here is the resource list i've been using:
- the tut: https://xon.sh/tutorial_xontrib.html#structure
- the template: https://github.com/xonsh/xontrib-template
- the source code: https://github.com/xonsh/xonsh/blob/main/xonsh/xontribs.py

In preface, i think my main critique is that there seems to be lots of "templating code / repos" for how to produce some `xontrib` but very little docs on the details of how to do such things manually 😉 

----
#### things i still don't grok:
- afaiu there are *2 main ways* to offer a `xontrib`:
  - through exposing it in some external package, as is most documented in the [structure docs](https://xon.sh/tutorial_xontrib.html#structure)
  - or, (presumably) through some other explicit loading (via `xontrib load -f` or the py api) from wtv namespaces are immediately available inside the `~/.xonshrc` environment?
    - why nothing in the docs mentions the `-f` flag is somewhat baffling, i had to read the source code to figure it out!
    - can you load such a `xontrib` without it being in an environment registered package? for eg. if you load some sibling module to your `rc.xsh` file?
    - what would be the best practice for file org if you are presumably hacking a `xontrib` this way to start, with the eventual intention of packaging it as per the `setuptools` methodology as is well documented?
- there seems to be a difference between `xontrib`s which are **auto-loaded** and those which are not but instead are expected to be **manually** loaded, normally via `xontrib load -f` (again, took me way to long to figure out the `-f` :joy:)
- there are 2 ways to offer `xontrib`s as 2 different file types: `.xsh` files and `.py` python modules but why would you use one over the other?
  - `.xsh` is some superset of python, so why ever offer a `.py`?
  - why would you put a `_load_xontrib_()` hook inside a `.xsh` module?
    - it seems as though you can access global vars like `aliases:  dict[str, Callable]` directly? is there some advantage to using it over accessing the runtime env that `.xsh` files seem to be run with?
  - if there are good reasons for using one file type for the other, they should probably be documented?
- if you **don't** register a `.xsh` file using the "special" `setuptools` params (as mentioned below), but the ext pkg which contains the `xontrib/` dir **is installed in the `xonsh` py env**, you can still load the `xontrib` using `xontrib load <name>` ?

---
### Premises (for things i hopefully *do* grok?):
assuming I actually understand how things work now i just want to assert what i think is true:

- depending on the file type, there are 2 different ways to register the `xontrib` using `setuptools`, **but this is only important for them to be autoloaded**:
  - for `.py` modules, they must be exposed by some pkg using something like,
  ```python
  entry_points={
      "xonsh.xontribs": [
            "xontrib_name = ns.path.to.xontrib_py_module",
        ],
  },
  ```
  - for `.xsh` modules/files, they (supposedly?) must be exposed using something like,
  ```python
      package_dir={
        'xontrib': 'xontrib',
    },
    package_data={
        'xontrib': ['*.xsh'],
    },
  ```
  yes, i understand you can also expose `.py` modules this way, but why would you? I guess just so that you can org all xontribs in one dir?
- if you want to offer a `xontrib` that is intentionally **not autoloaded** you only need to write what? nothing except for the file and add it to the python path such that you can load it with `xontrib load [-f]`?

### My critique:
- in the tut, there's a file-tree breakdown but no follow up to the comments beside each leaf node for how or why you'd use each potential `xontrib` entry and have it organized this way:
  - there should be small examples of the different ways in which you can offer `xontrib`s using both file types, 
  - auto-loading vs. no auto-loading, 
  - how (and presumably why) you'd use `_load_xontrib_()` hooks in (only?) `.py` modules.
    - yes i realize the obvious answer is that you're implementing the `xontrib` file as a `.py` mod and thus want to hook into the runtime loading, but this isn't really explicitly documented..
- there **definitely needs to be** documentation on the `xontrib load -f` flag for doing fully qualified module loading!
  - pretty sure the tut is borked without it for the example snippet after the file-tree section: `xontrib load myext mpl mypkg.show`
  
---
### in summary:
I'm more then happy to put up a PR to detail the tutorial with the examples that would address all of my own quiffs (presuming I have at least some of this figured out). I'm super interested to hear any explanation for the points I'm still unsure on; i think getting the details fully laid out in the docs will definitely help with more `xontrib`s being produced by new users in the future 🏄🏼

**Comment by goodboy** [comment]
On a somewhat related note. I think it would also be super great to have a small tut in the docs for how one might setup a user specific environment which can be kept isolated from the sys env, presuming the user would like to use said user env to be their main `xonsh` shell.

For ex, as far as i've gleaned, tools like `pyenv`, `xpip`/`pipx` and obviously some kind of special `xonsh` boot cmd could be used to accomplish this?

**Comment by anki-code** [comment]
I want to leave the link to xontrib template here https://github.com/xonsh/xontrib-template
